src/Compilers/VisualBasic/Portable/Binding/Binder_Lookup.vb

```diff
@@ -899,8 +918,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                     submission = submission.PreviousSubmission
                 Loop Until submission Is Nothing
 
-                If Not result.HasSymbol Then
-                    result.SetFrom(nonViable)
+                If result Is Nothing OrElse Not result.HasSymbol Then
+                    If result Is Nothing AndAlso nonViable.Kind <> LookupResultKind.Empty Then
```

Comment:
>If result Is Nothing AndAlso nonViable.Kind <> LookupResultKind.Empty Then [](start = 20, length = 74)

It feels fragile that we are basing the logic here on intimate knowledge of what SetFrom is going to do. One of the reasons that logic was extracted is to avoid duplicating it in multiple places and here (and in other similar places) we are reintroducing the duplication, making it unsafe to modify SetFrom's implementation. The same issue probably exists around other "merge" helpers that are surrounded with custom logic to pre-allocate/avoid-invocation logic.   
It feels like the code could be somewhat simpler and more allocations could probably be avoided if we convert all the affected SetFrom/"merge" helpers into extension methods that take "this" parameter ByRef and allocate on as needed bases inside them. Ideally, in this file we should only be left with null checks before dereferencing and no code that tries to predict what helpers are going to do.